api/src/main/java/org/infinispan/api/sync/Cache.java

```diff
@@ -0,0 +1,153 @@
+package org.infinispan.api.sync;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.infinispan.api.common.KeyValueEntry;
+import org.infinispan.api.common.events.KeyValueListener;
+import org.infinispan.api.configuration.CacheConfiguration;
+
+import io.smallrye.mutiny.Uni;
+
+
+/**
+ *
+ * @since 13.0
+ **/
+public interface Cache<K, V> {
+   /**
+    * Get the value of the Key if such exists
+    *
+    * @param key
+    * @return the value
+    */
+   V get(K key);
+
+   /**
+    * Insert the key/value pair. Returns the previous value if present.
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   V put(K key, V value);
+
+   /**
+    * Similar to {@link #put(Object, Object)} but does not return the previous value.
+    */
+   void set(K key, V value);
+
+   /**
+    * Save the key/value.
+    *
+    * @param key
+    * @param value
+    * @return true if the entry was put
+    */
+   boolean putIfAbsent(K key, V value);
+
+   /**
+    * Delete the key
+    *
+    * @param key
+    * @return true if the entry was removed
+    */
+   boolean remove(K key);
+
+   /**
+    * Retrieve all keys
+    *
+    * @return
+    */
+   Iterator<K> keys();
+
+   /**
+    * Retrieve all entries
+    *
+    * @return
+    */
+   Iterator<? extends KeyValueEntry<K, V>> entries();
+
+   /**
+    * Puts all entries
+    *
+    * @param entries
+    * @return Void
+    */
+   void putAll(Map<K, V> entries);
+
+   /**
+    * Estimate the size of the store
+    *
+    * @return Long, estimated size
+    */
+   long estimateSize();
+
+   /**
+    * Clear the store. If a concurrent operation puts data in the store the clear might not properly work.
+    */
+   void clear();
+
+   /**
+    * Executes the query and returns an {@link Iterable}
+    *
+    * @param query query String
+    * @return
+    */
+   <R> Iterable<R> find(String query);
```

Comment:
Given that I would expect the query to be fired by the invoking thread `Iterator<R>` seems applicable to me.